source/extensions/transport_sockets/tls/handshaker_impl.h

```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+#include "envoy/ssl/handshaker.h"
+#include "envoy/ssl/socket_state.h"
+
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Tls {
+
+// Class to abstract handshaking behavior.
+// Manages translation between SSL error codes and Network::PostIoAction
+// response enums, among other things.
+class HandshakerImpl : public Envoy::Ssl::Handshaker {
+public:
+  HandshakerImpl(bssl::UniquePtr<SSL> ssl) : ssl_(std::move(ssl)) {}
+
+  Network::PostIoAction doHandshake(Envoy::Ssl::SocketState& state) override;
+
+  void setCallbacks(Network::TransportSocketCallbacks& transport_socket_callbacks,
+                    Ssl::HandshakerCallbacks& handshaker_callbacks) override {
+    transport_socket_callbacks_ = &transport_socket_callbacks;
+    handshaker_callbacks_ = &handshaker_callbacks;
+  }
+
+  SSL* ssl() override { return ssl_.get(); }
```

Comment:
I think we're missing an important restriction on the use of this method: access to the SSL* object before handshake is complete should be forbidden.

I think this restriction clashes with use of this method in SslSocketInfo's constructor and possibly in a few other places.